dart/common/Composite.h

```diff
@@ -0,0 +1,212 @@
+
+  /// Check if this Composite is specialized for a specific type of Aspect.
+  ///
+  /// By default, this simply returns false.
+  template <class T>
+  static constexpr bool isSpecializedFor();
+
+  /// Check if this Composite requires this specific type of Aspect
+  template <class T>
+  bool requiresAspect() const;
+
+  /// Set the states of the aspects in this Composite based on the given
+  /// Composite::State. The states of any Aspect types that do not exist
+  /// within this composite will be ignored.
+  void setCompositeState(const State& newStates);
+
+  /// Get the states of the aspects inside of this Composite
+  State getCompositeState() const;
+
+  /// Fill outgoingStates with the states of the aspects inside this Composite
+  void copyCompositeStateTo(State& outgoingStates) const;
+
+  /// Set the properties of the aspects in this Composite based on the given
+  /// Composite::Properties. The properties of any Aspect types that do not
+  /// exist within this composite will be ignored.
+  void setCompositeProperties(const Properties& newProperties);
+
+  /// Get the properties of the aspects inside of this Composite
+  Properties getCompositeProperties() const;
+
+  /// Fill outgoingProperties with the properties of the aspects inside this
+  /// Composite
+  void copyCompositePropertiesTo(Properties& outgoingProperties) const;
+
+  /// Give this Composite a copy of each Aspect from otherComposite
+  void duplicateAspects(const Composite* fromComposite);
+
+  /// Make the Aspects of this Composite match the Aspects of otherComposite. Any
+  /// Aspects in this Composite which do not exist in otherComposite will be
+  /// erased.
+  void matchAspects(const Composite* otherComposite);
+
+protected:
+
+  template <class T> struct type { };
+
+  /// Add this Aspect to the Composite. This allows derived Composite types to
+  /// call the protected Aspect::setComposite function.
+  void addToComposite(Aspect* aspect);
+
+  /// Remove this Aspect from the Composite. This allows derived Composite types
+  /// to call the protected Aspect::loseComposite function.
+  void removeFromComposite(Aspect* aspect);
+
+  /// Non-templated version of set(const T*)
+  void _set(std::type_index type_idx, const Aspect* aspect);
+
+  /// Non-templated version of set(std::unqiue_ptr<T>&&)
+  void _set(std::type_index type_idx, std::unique_ptr<Aspect> aspect);
+
+  /// A map that relates the type of Aspect to its pointer
+  AspectMap mAspectMap;
+
+  /// A set containing type information for Aspects which are not allowed to
+  /// leave this composite.
+  RequiredAspectSet mRequiredAspects;
+};
+
+//==============================================================================
+/// Attach an arbitrary number of Aspects to the specified Composite type.
+// TODO(MXG): Consider putting this functionality into the Composite class
+// itself. Also consider allowing the user to specify arguments for the
+// constructors of the Aspects.
```

Comment:
Maybe we can put this function into `Composite` as:

```cpp
// header
template <class... Aspects>
  void createAspects();

// source
//==============================================================================
template <class... Aspects>
struct CreateAspectsImpl;

//==============================================================================
template <class FirstAspect, class... RestAspects>
struct CreateAspectsImpl<FirstAspect, RestAspects...>
{
  static void create(Composite* composite)
  {
    composite->createAspect<FirstAspect>();

    CreateAspectsImpl<RestAspects...>::create(composite);
  }
};

//==============================================================================
template <>
struct CreateAspectsImpl<>
{
  static void create(Composite* /*composite*/) {}
};

//==============================================================================
template <class... Aspects>
void Composite::createAspects()
{
  CreateAspectsImpl<Aspects...>::create(this);
}
```

Passing arguments for each `Aspect` might be tricky. One possible solution I can come up with now is using `std::tuple` with [this workaround](http://stackoverflow.com/questions/7858817/unpacking-a-tuple-to-call-a-matching-function-pointer) to unpack `std::tuple` beck.